### ISSUE TITLE
Functions: Misspelling the keyword function, is not caught. Rather, code is directly executed, changing the main() entry point.

### ISSUE THREAD

**Issue by gh-greg** [koalaman/shellcheck#1528]
#### For new checks and feature suggestions
SUMMARY: Functions: Misspelling the keyword function, is not caught. Rather, code is directly executed, changing the main() entry point.
[x ] shellcheck.net currently gives no useful warnings about this. "No issues detected!"
[x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#! /bin/bash

functions  fn_some_dangerous_code_IMproper_spelling
{
    echo -e "----------\nINFO: Start of some dangerous code, below.\nThis code is intended to be constrined to run within a function call." 
    git reset --hard
    git checkout master
    git pull
}

function  fn_some_dangerous_code_PROPER_spelling
{
    echo -e "----------\nINFO: Start of some dangerous code, below.\nThis code is intended to be constrined to run within a function call."    
    git reset --hard
    git checkout master
    git pull
}

cat << END_main
----------
INFO: Starting the main() code.
Test Case: Shouldn't shellcheck, flag this as a MAL-formed function shown below ?
----------
function_s  fn_some_dangerous_code
{
    :FUNCTION_CODE_GOES_HERE
}
Test Case: Shouldn't we get a Parse-Time error and abort ?
Test Case: Instead of aborting the sample dangerous code is run out of order. Ran directly. Skipping the the MAL-formed / misspelled function
END_main

```

#### Here's what shellcheck currently says:
c:\gregpf\python\aaa>> checkshell  shellcheck_bug_01.sh

c:\gregpf\python\aaa>> powershell shellcheck  --check-sourced  --color=never  --shell=bash    --wiki-link-count=20  --external-sources  "she
llcheck_bug_01.sh"
c:\gregpf\python\aaa>>

#### Here's what I wanted or expected to see:
```sh
functions  fn_some_dangerous_code_improper_spelling
{
^-- SC999999: Function definition was not preceded by the keyword "function"

```

**Comment by Gandalf-** [comment]
Unless we checked for common misspellings of `function`, I'm not sure this is possible. Since the code in your example is perfectly valid.

Another example
```
funktion args to command named funktion

{
  ... unrelated shell block ...
}
```

This seems like a task that may be better suited to your text editor, because it won't highlight misspelled keywords.

**Comment by gh-greg** [comment]
(1) @Gandalf- : Good comment. I think net-net, you are right. (Darn wise wizard !)
(2) Run the code below. I cannot find a case where ShellCheck could do more, than what Bash already does.
(3) I'm closing the ticket.  THANKS to the author(s) of ShellCheck.
--------------------- 
`#! /bin/bash

echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t "

fn_style_1_no_keyword_function ()
{
    echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t ___ "
}

function fn_style_2a_keyword_function_no_parens 
{
    echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t ___ "
}

function fn_style_2a_keyword_function_yes_parens  ()
{
    echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t ___ "
}

### functions fn_style_2a_keyword_function_yes_parens_mispelled  ()  ### syntax error near unexpected token `('
### {
###     echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t ___ "
### }

echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t Starting main() ..."
fn_style_1_no_keyword_function

echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t "
fn_style_2a_keyword_function_no_parens

echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t "
fn_style_2a_keyword_function_yes_parens

cat << INFO_INFO_INFO
##################################################
https://www.gnu.org/software/bash/manual/html_node/Shell-Functions.html

Functions are declared using this syntax:

name () compound-command [ redirections ]
((OR))
function name [()] compound-command [ redirections ]

The reserved word function is optional.
If the function reserved word is supplied, the parentheses are optional.
The body of the function is the compound command compound-command (see Compound Commands).
That command is usually a list enclosed between { and }
, but may be any compound command listed above, with one exception:
If the function reserved word is used, but the parentheses are not supplied, the braces are required.

https://www.gnu.org/software/bash/manual/html_node/Compound-Commands.html#Compound-Commands
https://www.gnu.org/software/bash/manual/html_node/Command-Grouping.html#Command-Grouping

3.2.4.3 Grouping Commands
Bash provides two ways to group a list of commands to be executed as a unit.
When commands are grouped, redirections may be applied to the entire command list.
For example, the output of all the commands in the list may be redirected to a single stream.

()
( list )
Placing a list of commands between parenthesescauses a subshell environment to be created 
(see Command Execution Environment), and each of the commands in list to be executed in that subshell.
Since the list is executed in a subshell, variable assignments do not remain in effect after the subshell completes.

{}
{ list; }
Placing a list of commands between curly braces causes the list to be executed in the current shell context.
No subshell is created.
The semicolon (or newline) following list is required.

In addition to the creation of a subshell, there is a subtle difference between these two constructs due to historical reasons.
The braces are reserved words, so they must be separated from the list by blanks or other shell metacharacters.
The parentheses are operators, and are recognized as separate tokens by the shell even if they are not separated from the list by whitespace.

The exit status of both of these constructs is the exit status of list.
INFO_INFO_INFO

echo -e "LINE: $LINENO:\t${FUNCNAME[0]}:\t Leaving the entire script " `
